pkg/server/forward.go

```diff
@@ -51,68 +53,74 @@ func init() {
 	prometheus.MustRegister(overwrittenTimestamps)
 }
 
-type Store struct {
-	next   store.Store
-	url    *url.URL
-	client *http.Client
-	logger log.Logger
-}
+// ForwardHandler gets a request containing metric families and
+// converts it to a remote write request forwarding it to the upstream at fowardURL.
+func ForwardHandler(logger log.Logger, forwardURL *url.URL) http.HandlerFunc {
+	client := http.Client{}
 
-func New(logger log.Logger, url *url.URL, next store.Store) *Store {
-	return &Store{
-		next:   next,
-		url:    url,
-		client: &http.Client{},
-		logger: log.With(logger, "component", "store/forward"),
-	}
-}
+	return func(w http.ResponseWriter, r *http.Request) {
+		partitionKey, ok := PartitionFromContext(r.Context())
+		if !ok {
+			http.Error(w, "failed to retreive partitionKey", http.StatusInternalServerError)
+			return
+		}
 
-func (s *Store) ReadMetrics(ctx context.Context, minTimestampMs int64) ([]*store.PartitionedMetrics, error) {
-	return s.next.ReadMetrics(ctx, minTimestampMs)
-}
+		decoder := expfmt.NewDecoder(r.Body, expfmt.ResponseFormat(r.Header))
+		defer r.Body.Close()
+
+		families := make([]*clientmodel.MetricFamily, 0, 100)
+		for {
+			family := &clientmodel.MetricFamily{}
+			if err := decoder.Decode(family); err != nil {
+				if err == io.EOF {
+					break
+				}
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
 
-func (s *Store) WriteMetrics(ctx context.Context, p *store.PartitionedMetrics) error {
-	if p == nil {
-		return nil
-	}
+			families = append(families, family)
+		}
+		families = metricfamily.Pack(families)
 
-	err := func() error {
-		timeseries, err := convertToTimeseries(p, time.Now())
+		timeseries, err := convertToTimeseries(&PartitionedMetrics{PartitionKey: partitionKey, Families: families}, time.Now())
 		if err != nil {
-			return err
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
 		}
 
 		if len(timeseries) == 0 {
-			level.Info(s.logger).Log("msg", "no time series to forward to receive endpoint")
-			return nil
+			level.Info(logger).Log("msg", "no time series to forward to receive endpoint")
```

Comment:
Should we also respond with an HTTP error here?

Reply:
It was a transient error before. We simply ignored those requests and returned 200. I don't really care, honestly.